addons/web/static/src/js/views/abstract_renderer.js

```diff
@@ -111,6 +127,56 @@ return mvc.Renderer.extend({
     _render: function () {
         return Promise.resolve();
     },
+    /**
+     * Assigns a new state to the renderer if not false.
+     *
+     * @private
+     * @param {any} [state=false]
+     */
+    _setState(state = false) {
+        if (state !== false) {
+            this.state = state;
+        }
+    },
+    /**
+     * Suppresses tabindex property on any focusable element located inside
+     * root elements defined in the `this.sampleDataTargets` object. This is
+     * done to avoid focusing sample or any of their widget.
+     *
+     * @private
+     */
+    _toggleSampleDataModifiers() {
+        // Remove previous modifiers
+        if (this.removeSampleModifiers) {
+            this.removeSampleModifiers();
+        }
+        if (this.state.isSample) {
+            const rootEls = [];
+            for (const selector of this.sampleDataTargets) {
+                rootEls.push(...this.el.querySelectorAll(':scope ' + selector));
+            }
+            const focusableEls = new Set(rootEls);
+            for (const root of rootEls) {
+                for (const focusableEl of root.querySelectorAll(FOCUSABLE_ELEMENTS)) {
+                    focusableEls.add(focusableEl);
+                }
+            }
+            // Suppress tab indices
+            const originalTabindices = new Map();
+            for (const focusableEl of focusableEls) {
+                originalTabindices.set(focusableEl, focusableEl.tabindex);
+                focusableEl.setAttribute('tabindex', -1);
+            }
+            this.removeSampleModifiers = () => {
+                // Restore tab indices
+                for (const focusableEl of focusableEls) {
+                    focusableEl.setAttribute('tabindex', originalTabindices.get(focusableEl));
+                }
+            };
+        } else {
+            this.removeSampleModifiers = null;
+        }
+    },
```

Comment:
Do you really need to store them/restore them later? Can't we rely on the fact that _render completely replaces the current DOM by a new DOM in which tab indices will be reset?

Reply:
Indeed, the method is systematically called after `_render`